cupy/core/core.pyx

```diff
@@ -1717,51 +1719,177 @@ cpdef ndarray array(obj, dtype=None, bint copy=True, order='K',
     elif hasattr(obj, '__cuda_array_interface__'):
         return array(_convert_object_with_cuda_array_interface(obj),
                      dtype, copy, order, subok, ndmin)
+    else:  # obj is sequence, numpy array, scalar or the other type of object
+        shape, elem_dtype, head = _get_concat_shape(obj)
+        if shape is not None and shape[-1] != 0:
+            # obj is a non-empty sequence of ndarrays which share same shape
+            # and dtype
+
+            # resulting array is C order unless 'F' is explicitly specified
+            # (i.e., it ignores order of element arrays in the sequence)
+            order = (
+                'F'
+                if order is not None and len(order) >= 1 and order[0] in 'Ff'
+                else 'C')
+            if elem_dtype.char not in '?bhilqBHILQefdFD':
+                raise ValueError('Unsupported dtype %s' % elem_dtype)
+            ndim = len(shape)
+            if ndmin > ndim:
+                shape = (1,) * (ndmin - ndim) + shape
+
+            if dtype is None:
+                dtype = elem_dtype
+            if isinstance(head, numpy.ndarray):  # obj is Seq[numpy.ndarray]
+                a = _send_numpy_array_list_to_gpu(
+                    obj, dtype, shape, order, ndmin)
+            elif isinstance(head, ndarray):  # obj is Seq[cupy.ndarray]
+                a = (_manipulation.concatenate_method(_flatten_list(obj), 0)
+                                  .reshape(shape)
+                                  .astype(dtype, order=order, copy=False))
+            else:  # should not be reached here
+                raise ValueError(
+                    "The elements of obj are unsupported type(s)")
+        else:
+            # obj is:
+            # - numpy array
+            # - scalar or sequence of scalar
+            # - empty sequence or sequence with elements whose shapes or
+            #   dtypes are unmatched
+            # - the other type of object
+
+            # fallback to numpy array and send it to GPU
+            a = _send_object_to_gpu(obj, dtype, order, ndmin)
+
+    return a
+
+
+cdef tuple _get_concat_shape(object obj):
+    # Returns a tuple of the following:
+    # 1. concatenated shape if it can be converted to a single CuPy array by
+    #    just concatenating it (i.e., the object is a (nested) sequence only
+    #    which contains NumPy/CuPy array(s) with same shape and dtype).
+    #    Returns None otherwise.
+    # 2. dtype if the object is an array
+    # 3. a first element in the object
```

Comment:
How about using "item" instead of "element"?
"Element" more commonly refers to a scalar in arrays.